File: atcoder/ABC/461/c_ans_official.cpp
// g++ -D LOCAL c_ans.cpp
// ./a.out

#include <bits/stdc++.h>
using namespace std;

int main(void){
// ==========================================
    // 提出時はここから下のブロックを削除・コメントアウトする
    // ローカル環境でのみ実行されるブロック
#ifdef LOCAL
    string test_input = R"(
5 3 3
1 30
1 40
1 50
2 10
3 20
)";
    stringstream ss(test_input);
    cin.rdbuf(ss.rdbuf()); // cin の読み込み先を test_input に変更
#endif
// ==========================================

    int n, k, m;
    cin >> n >> k >> m;

    // t[color] に、その色の宝石の価値をまとめる
    // Pythonの t = [[] for _ in range(N + 1)] に相当
    vector<vector<long long>> t(n + 1);

    for (int i = 0; i < n; i++) {
        int c;
        long long v;
        cin >> c >> v;
        t[c].push_back(v);
    }

    // top  : 各色の代表宝石
    // tail : 代表以外の宝石
    vector<long long> top;
    vector<long long> tail;

    // 色ごとに処理
    for (int i = 1; i <= n; i++) {
        // その色の宝石が1つもないならスキップ
        if (t[i].empty()) {
            continue;
        }

        // 価値が高い順（降順）に並べる
        sort(t[i].rbegin(), t[i].rend());

        // 一番価値が高いものをその色の代表にする
        top.push_back(t[i][0]);

        // 残りは代表以外の候補に入れる (Pythonの tail += values[1:])
        for (size_t j = 1; j < t[i].size(); j++) {
            tail.push_back(t[i][j]);
        }
    }

    // 代表宝石を価値が高い順に並べる
    sort(top.rbegin(), top.rend());

    long long answer = 0;

    // 代表のうち、上位M個は必ず選ぶ
    for (int i = 0; i < m; i++) {
        answer += top[i];
    }

    // M個に入らなかった代表は、普通の候補(tail)に戻す
    for (size_t i = m; i < top.size(); i++) {
        tail.push_back(top[i]);
    }

    // 残り候補を価値が高い順に並べる
    sort(tail.rbegin(), tail.rend());

    // 代表M個を選んだので、あとK-M個をtailから選ぶ
    for (int i = 0; i < k - m; i++) {
        answer += tail[i];
    }

    cout << answer << endl;

    return 0;
}

File: atcoder/ABC/461/c_ans_official.py
import io
import sys
from collections import Counter

# 下記に標準入力を記載
_INPUT = """\
5 3 3
1 30
1 40
1 50
2 10
3 20
"""
sys.stdin = io.StringIO(_INPUT)

# ここからコードを記載
N, K, M = map(int, input().split())

# t[color] に，その色の宝石の価値をまとめる
t = [[] for _ in range(N + 1)]

# 以下のように配列をあたかも数値のキーを持つ辞書の様に利用する
# t[1] = 色1の宝石価値リスト
# t[2] = 色2の宝石価値リスト
# t[3] = 色3の宝石価値リスト
for _ in range(N):
    C, V = map(int, input().split())
    t[C].append(V)

# top  : 各色の代表宝石，つまりその色で一番価値が高い宝石
# tail : 代表以外の宝石
top = []
tail = []


# ソートした後の箱（例：[50, 40, 30]）から、
# 一番大きい代表（values[0] つまり 50）を取り除いた残りすべて（values[1:] つまり [40, 30]）を、
# 一発で tail（落選者リスト）に合流させています。

for values in t:
    # その色の宝石が1つもないならスキップ
    if len(values) == 0:
        continue

    # 価値が高い順に並べる
    values.sort(reverse=True)

    # 一番価値が高いものをその色の代表にする
    top.append(values[0])

    # 残りは代表以外の候補に入れる
    tail += values[1:]

# 代表宝石を価値が高い順に並べる
top.sort(reverse=True)

# 代表のうち，上位M個は必ず選ぶ
selected_top = top[:M]

# M個に入らなかった代表は，普通の候補に戻す
tail += top[M:]

# 残り候補を価値が高い順に並べる
tail.sort(reverse=True)

# 代表M個を選んだので，あとK-M個をtailから選べばよい
answer = sum(selected_top) + sum(tail[:K - M])

print(answer)